File: benchmarks/cpp/map_lookup.cpp
// BENCH_MAP_SIZE=5000 BENCH_MAP_LOOKUPS=50000
#include <cstdlib>
#include <iostream>

static int key_at(int i) { return i * 3 + 7; }
static int value_at(int i) { return i * 31 + 13; }

static int lookup(int keys_size, int target) {
    for (int i = 0; i < keys_size; ++i) {
        if (key_at(i) == target) return value_at(i);
    }
    return -1;
}

static long long lcg_next(long long state) {
    long long s = state * 1103515245LL + 12345LL;
    return s < 0 ? -s : s;
}

static long long map_workload(int map_size, int lookups, int seed) {
    long long sum = 0;
    long long state = seed;
    for (int n = 0; n < lookups; ++n) {
        state = lcg_next(state);
        int idx = static_cast<int>(state % map_size);
        sum += lookup(map_size, key_at(idx));
    }
    return sum;
}

int main() {
    std::cout << map_workload(5000, 50000, 42) << '\n';
    return 0;
}

File: benchmarks/rust/src/bin/map_lookup.rs
fn key_at(i: i32) -> i32 {
    i * 3 + 7
}

fn value_at(i: i32) -> i32 {
    i * 31 + 13
}

fn lookup(keys_size: i32, target: i32) -> i32 {
    for i in 0..keys_size {
        if key_at(i) == target {
            return value_at(i);
        }
    }
    -1
}

fn lcg_next(state: i64) -> i64 {
    let s = state * 1_103_515_245 + 12_345;
    if s < 0 { -s } else { s }
}

fn map_workload(map_size: i32, lookups: i32, seed: i32) -> i64 {
    let mut sum = 0_i64;
    let mut state = seed as i64;
    for _ in 0..lookups {
        state = lcg_next(state);
        let idx = (state % map_size as i64) as i32;
        sum += lookup(map_size, key_at(idx)) as i64;
    }
    sum
}

fn main() {
    println!("{}", map_workload(5000, 50_000, 42));
}
